### PR TITLE
Shared mem

### PR DESCRIPTION
**Pull request by MattJaniszewski**
I ran into a limit when the FIFO pipe would get full (> ~62k of data), and cause the child processes to block until the pipe was cleared (e.g., by running the `cat` command in the appropriate FIFO handle on the file system). The use of the `shmop_` functions seemed like a good alternative.

In the PR, I added a unit test which re-creates the problem that kept the original version from working on a typical Linux system.

One thing I will note is that while I did all the due-diligence I could with the `shmop_` functions, I had to end up using `@shmop_open` with the error suppressor operator twice, to test for the presence for an existing memory segment. It's not ideal, but seemed like the only way.

A decent way to test that there are no left-over open memory segments is to run `ipcs -m` on the Linux system. You can then pass any offending `shmid` value over to the `ipcrm` binary. My testing as of yet indicated no issues of this kind, but it's theoretically possible it can happen in some edge-cases.

### CODE REVIEW

**Comment by seyfer** [comment]
:+1:

**Comment by githoober** [comment]
@MattJaniszewski: :+1:

File: src/Spork/SharedMem.php [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        $this->ppid = $ppid;
+        $this->signal = $signal;
+    }
+
+    /**
+     * Reads all messages from shared memory.
+     *
+     * @return array An array of messages
+     */
+    public function receiveMessages()
```
Comment (githoober) [inline src/Spork/SharedMem.php]: In what cases will the result value contain more then one message?

File: src/Spork/ProcessManager.php [hunk 2]
```diff
@@ -99,22 +99,22 @@ public function fork($callable)
...
 
-            // setup the fifo (blocks until parent connects)
-            $fifo = new Fifo(null, $this->signal);
+            // setup the shared memory
+            $sharedMem = new SharedMem(null, $this->signal);
             $message = new ExitMessage();
 
             // phone home on shutdown
-            register_shutdown_function(function() use(&$fifo, &$message) {
+            register_shutdown_function(function () use (&$sharedMem, &$message) {
```
Comment (githoober) [inline src/Spork/ProcessManager.php]: Is there a reason to pass two parameters by a reference?

File: src/Spork/SharedMem.php (hunk 1 above)
Comment (MattJaniszewski) [inline src/Spork/SharedMem.php]: Thanks Oleg. I took a look again at this code, and realized that this method wasn't actually working as advertised. I think that I fixed the logic now, so take another look when you can.

File: src/Spork/ProcessManager.php (hunk 2 above)
Comment (MattJaniszewski) [inline src/Spork/ProcessManager.php]: I removed them, since it appears to be not necessary in my case anymore.

**Comment by kriswallsmith** [comment]
Would you mind removing the PSR-2 commit (or moving it to another pull request) so that this pull request is limited to the shared memory change?

**Comment by MattJaniszewski** [comment]
I rebased this branch, and removed all the un-related modifications. I created separate branches for them, and I'll create the pull requests later (so they will end up being fast-forward merges).

**Pull request merged**